File: ruby-service/lib/routing.rb
# frozen_string_literal: true

module Routing
  EXCHANGE = 'music.events'

  # Slugify a value the SAME way the TypeScript service does (see ts-service/src/routing.ts):
  # lowercase, replace any run of non-alphanumeric characters with a single '-',
  # and strip leading/trailing '-'. This keeps the routing key free of '.' so it
  # never breaks AMQP topic word boundaries.
  def self.slug(value)
    value.to_s.downcase.gsub(/[^a-z0-9]+/, '-').gsub(/\A-+|-+\z/, '')
  end

  # Build the topic routing key for an event: events.<city>.<genre>
  def self.routing_key_for(event)
    city = slug(event.dig('location', 'city'))
    genre = slug(event['genre'])
    "events.#{city}.#{genre}"
  end
end

File: ts-service/src/routing.ts
export const EXCHANGE = 'music.events';

export interface MusicEvent {
  id: string;
  title: string;
  artist: string;
  genre: string;
  location: { city: string; venue?: string; country: string };
  startsAt: string;
  endsAt?: string;
  price?: { amount: number; currency: string };
  submittedBy: string;
  submittedAt: string;
  [key: string]: unknown;
}

/**
 * Slugify a value the SAME way the Ruby service does (see ruby-service/lib/routing.rb):
 * lowercase, collapse any run of non-alphanumerics to a single '-', strip ends.
 * Keeps the routing key free of '.' so it never breaks AMQP topic word boundaries.
 */
export function slug(value: unknown): string {
  return String(value ?? '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '');
}

/** Build the topic routing key for an event: events.<city>.<genre> */
export function routingKeyFor(event: Partial<MusicEvent>): string {
  const city = slug(event.location?.city);
  const genre = slug(event.genre);
  return `events.${city}.${genre}`;
}
